### PR TITLE
OpenShift SDN: Changed ovs.Transaction from pseudo to real atomic transaction

### PR DESCRIPTION
**Pull request by pravisankar**
- Leverages ovs.bundle() to perform atomic transactions
- Now ovs.Transaction interface has AddFlow(), DeleteFlows() and Commit() methods
  General usage:
  otx := ovs.NewTransaction()
  otx.AddFlow(flow1) // No execution, only caches flow context
  ...
  otx.DeleteFlows(flowN) // No execution, only caches flow context
  ...
  err := otx.Commit() // Executes all cached flows as single atomic transaction
- With this change, most of the operations in ovs controller like setup,
  addHostSubnet, addService, etc. has only one ovs bundle call. So there
  won't be partial commited changes and it reduces the downtime during
  watch resync events.

### CODE REVIEW

**Comment by pravisankar** [comment]
@openshift/sig-networking PTAL

File: pkg/util/ovs/ovs.go [hunk 1]
```diff
@@ -138,13 +142,23 @@ func New(execer exec.Interface, bridge string, minVersion string) (Interface, er
...
-	glog.V(4).Infof("Executing: %s %s", cmd, strings.Join(args, " "))
 
-	output, err := ovsif.execer.Command(cmd, args...).CombinedOutput()
+	kcmd := ovsif.execer.Command(cmd, args...)
+	if stdinArgs != nil {
+		stdinString := strings.Join(stdinArgs, "\n")
+		stdin := bytes.NewBufferString(stdinString)
+		kcmd.SetStdin(stdin)
+
+		glog.V(4).Infof("Executing: %s %s |\n%s", cmd, strings.Join(args, " "), stdinString)
```
Comment (danwinship) [inline pkg/util/ovs/ovs.go]: The use of `|` there is weird: it makes it look like you're piping the output of ovs-ofctl to the string.
Actually, just changing `|` to `<<` should make it clear what's going on, shell-syntax-wise, I think?

File: pkg/util/ovs/fake_ovs.go [hunk 2]
```diff
@@ -248,6 +250,52 @@ func (tx *ovsFakeTx) EndTransaction() error {
 	return err
 }
 
+func getFlowRule(bundleFlow string) (string, bool, error) {
+	var parts []string
+	isAdd := false
+	addPrefix := "flow add"
+	delPrefix := "flow delete"
+
+	if strings.Contains(bundleFlow, addPrefix) {
```
Comment (danwinship) [inline pkg/util/ovs/fake_ovs.go]: `strings.HasPrefix`

File: pkg/util/ovs/fake_ovs.go [hunk 3]
```diff
@@ -283,16 +284,12 @@ func (fake *ovsFake) Bundle(bundleFlows []string) error {
 		}
 
 		if isAdd {
-			tx.AddFlow(flow)
+			fake.addFlowHelper(flow)
```
Comment (danwinship) [inline pkg/util/ovs/fake_ovs.go]: you need to check the error return here

**Comment by pravisankar** [comment]
@danwinship  Updated, PTAL

File: pkg/util/ovs/ovs.go [hunk 4]
```diff
@@ -106,7 +107,7 @@ type ovsExec struct {
 }
 
 // New returns a new ovs.Interface
-func New(execer exec.Interface, bridge string, minVersion string) (Interface, error) {
+func New(execer exec.Interface, bridge string, minVersion string) (*ovsExec, error) {
```
Comment (danwinship) [inline pkg/util/ovs/ovs.go]: Hm... isn't that bad style? Maybe you could just have TestBundle() cast it from Interface to *ovsExec.

Alternatively, after the next commit, TestBundle() doesn't really do anything beyond what TestTransactionSuccess() and TestTransactionFailure() do, so you could just drop TestBundle() (and maybe add a TestEmptyTransaction() in the next commit if you want to explicitly test that case too.)

File: pkg/network/node/ovscontroller.go [hunk 5]
```diff
@@ -496,7 +491,7 @@ func (oc *ovsController) UpdateEgressNetworkPolicyRules(policies []networkapi.Eg
 		otx.DeleteFlows("table=101, reg0=%d, cookie=1/1", vnid)
```
Comment (danwinship) [inline pkg/network/node/ovscontroller.go]: UpdateEgressNetworkPolicyRules no longer needs to do the "temporarily drop all outgoing traffic to avoid race conditions" thing.

You can also revert the changes from #19346 since they're no longer needed

File: pkg/util/ovs/ovs_test.go [hunk 6]
```diff
@@ -53,8 +53,7 @@ func ensureTestResults(t *testing.T, fexec *fakeexec.FakeExec) {
 
 func TestTransactionSuccess(t *testing.T) {
 	fexec := normalSetup()
-	addTestResult(t, fexec, "ovs-ofctl -O OpenFlow13 add-flow br0 flow1", "", nil)
-	addTestResult(t, fexec, "ovs-ofctl -O OpenFlow13 add-flow br0 flow2", "", nil)
+	addTestResult(t, fexec, "ovs-ofctl -O OpenFlow13 bundle br0 -", "", nil)
```
Comment (danwinship) [inline pkg/util/ovs/ovs_test.go]: (It would be nice if we could test the stdin value too...)

File: pkg/util/ovs/fake_ovs.go [hunk 7]
```diff
@@ -161,12 +163,12 @@ func (fake *ovsFake) Clear(table, record string, columns ...string) error {
...
 type ovsFakeTx struct {
-	fake *ovsFake
-	err  error
+	fake  *ovsFake
+	flows []string
 }
 
 func (fake *ovsFake) NewTransaction() Transaction {
-	return &ovsFakeTx{fake: fake, err: fake.ensureExists()}
+	return &ovsFakeTx{fake: fake, flows: []string{}}
```
Comment (danwinship) [inline pkg/util/ovs/fake_ovs.go]: Note that there's a mild behavior change here: it used to check that you had called `ovsif.AddBridge()` on the `ovsFake`, but now it doesn't. (We apparently don't test this case in `fake_ovs_test.go`.) Anyway, you could just move the `fake.ensureExists()` call to Commit

File: pkg/util/ovs/fake_ovs.go [hunk 8]
```diff
@@ -200,52 +202,78 @@ func fixFlowFields(flow *OvsFlow) {
...
-	tx.fake.flows = newFlows
+	fake.flows = newFlows
+	return nil
 }
 
-func (tx *ovsFakeTx) EndTransaction() error {
-	err := tx.err
-	tx.err = nil
-	return err
+func (tx *ovsFakeTx) Commit() error {
```
Comment (danwinship) [inline pkg/util/ovs/fake_ovs.go]: This doesn't actually implement transactional behavior: if you do

    tx.AddFlow("table=10, actions=drop")
    tx.AddFlow("@!#$!@#$!@#$")
    tx.Commit()

then it will return an error saying that the second flow couldn't be parsed, but the `"table=10, actions=drop"` flow will still have been added.

There should probably be a test of transactionality in `fake_ovs_test.go`

**Comment by pravisankar** [comment]
@danwinship Addressed all your comments, PTAL

**Comment by knobunc** [comment]
Looks good to me.

@dcbw can you take a look please?
